docs: add common patterns section to readme

imbolc opened the pull request.

jonahlund commented:
Thanks! Theres also one great pattern that i think alot of people will appreciate:

- **Zero cost concatenation with tuples**

Since `vy` is heavily tuple based, you can actually use tuples for zero cost concats where you would otherwise use format!() or such, like for example:

```rust
let id = 123;

a!(
  href = format!("/products/{id}")
)
```
Can actually be written like:
```rust
let id = 123;

a!(
  href = ("/products/", id)
)
```

imbolc commented:
> Perhaps this is already obvious?

At least I found it instinctively as Maud allows this :)

Actually I'm having an issue with the examples as docs have stricter lifetime rules. E.g. original function that works in normal code looked like this:

```rust
fn patterns(items: &[&str], condition: bool, maybe_some: Option<&str>) -> impl IntoHtml {
    (
        // for loop
        items.iter().map(|item| li!(*item)),
        // if-then-else
        condition.then_some(i!("condition is met")),
        (!condition).then_some(b!("condition is NOT met")),
        // if-let-else
        maybe_some.map(|inner| i!(inner)),
        maybe_some.is_none().then_some(b!("none")),
    )
}
```

I didn't want to scare potential users with explicit lifetimes, which they won't really need in their code, so I changed it to use owned types. And then I removed the else branch of if-let as it would require a clone with the owned string.

It seems that the 2024 edition doesn't have this lifetime issue, but I found you consciously decided not to use it: [#6765807a171db7a0a57b4e587827e4c3ba7157f2](https://github.com/jonahlund/vy/commit/6765807a171db7a0a57b4e587827e4c3ba7157f2). I wonder if the reasons to switch back to 2021 are still relevant?

jonahlund commented:
> It seems that the 2024 edition doesn't have this lifetime issue, but I found you consciously decided not to use it: [#6765807a171db7a0a57b4e587827e4c3ba7157f2](https://github.com/jonahlund/vy/commit/6765807a171db7a0a57b4e587827e4c3ba7157f2). I wonder if the reasons to switch back to 2021 are still relevant?

Yeah i did remember switching edition because of some lifetime crux back then, i will take a look at it again
